### DISCUSSION TITLE
Connecting to clickhouse db with mysql protocol

### DISCUSSION THREAD

**Discussion by mxfwang** [sfu-db/connector-x#165] [Q&A] [answered]
Hi, I'm trying to read from a clickhouse database and got this error

> RuntimeError: MySqlError { ERROR 48 (00000): Code: 48. DB::Exception: Command [ERRFMT] is not implemented. (NOT_IMPLEMENTED) (version 21.11.3.6 (official build)) }

I'm running very simple queries like `SELECT 1`

I'm not sure if this error is from clickhouse, or connectorx?

My setup is also very simple. I spinned up a local clickhouse on my machine and enabled mysql port 9004 like described in [this page](https://clickhouse.com/docs/en/interfaces/mysql/), exposed the port, then did this

```
import connectorx as cx
url = "mysql://default:@<hostname>:9004/"
query = "SELECT 1"
move = cx.read_sql(url, query)
```

**Comment by wangxiaoying** [comment, marked as answer]
Hi @mxfwang , can you add `protocol= text` in `read_sql` function and try again? Clickhouse only support text protocol and we use binary protocol by default in ConnectorX. More info can be found [here](https://github.com/sfu-db/connector-x/blob/main/Types.md#mysql-clickhouse).

**Reply by mxfwang** [reply]
That was it. Thank you!

**Comment by jeeinn** [comment]
if with PDO， how to config?

**Reply by wangxiaoying** [reply]
We haven't test on it. If they share the same protocol, can you try the method above? Or setting the schema to clickhouse like `clickhouse://username:password@host:port/db`? Internally we will transfer it to url = `mysql://username:password@host:port/db` and protocol=text.

**Reply by am000k** [reply]
Hi, I am also trying to test the connector with clickhouse 

import connectorx as cx
url = "mysql://default:@hostname:9004/"
query = "SELECT 1"
move = cx.read_sql(url, query, protocol="text")

and I get the error:
RuntimeError: MySqlError { ERROR 48 (00000): Code: 48. DB::Exception: Command [ERRFMT] is not implemented. (NOT_IMPLEMENTED) (version 22.1.3.7 (official build)) }

**Reply by wangxiaoying** [reply]
Hi @am000k , previously we updated the schema fetching using the prepared statement for mysql protocol and it turns out clickhouse's mysql interface does not support prepared queries. We are now adding the original limit1 method to get the schema when prepared statement is not supported and will release a new alpha version. Will let you know when it's ready. In the meanwhile, versions <=0.2.1 should works for clickhouse.

**Reply by wangxiaoying** [reply]
@am000k the new alpha version `0.2.4a7` is ready. Can you try whether it works?

**Reply by am000k** [reply]
Hi wangxiaoying, the alpha version works fine.
Thanks.
